Thunder-SVM: Extremely Slow MPI_finalize

Opening this by request of Servesh ...

Thunder-SVM is longer completing in the expected amount of time using the current Aurora MPICH.  The problem has been isolated to the default enablement of XpMem.

Thunder-SVM uses broadcasts to distribute a large input set.   The dominant communication is a series of 100,000 broadcasts, each of ~32k bytes.  Each message buffer is unique.   The result of this is twofold:  the broacast phase itself is slower with XpMem enabled.  However, what is really negatively impacting the jobs is the teardown of the > 900,000 shared memory VMAs during MPI_finalize.  There must be some heavy locking in the implementation of xpmem_detach (kernel module).

It is understood that simply disabling XpMem will improve Thunder-SVM.  This is simply documenting that the current defaults are harmful to a SOW app and could be impacting other applications as well.

Here is a simplified reproducer:

```
#define _GNU_SOURCE
#include <mpi.h>
#include <omp.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sched.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <errno.h>



static unsigned long *alloc_msg_buffer(unsigned long size, int rank, int initialize);
static void dump_self_maps(void);

int main(int argc, char *argv[])
{

  int  numiters = 1, reuse_buffer = 0, verbose = 0, dump_maps = 0;
  unsigned long bcast_length = 0;
  int  numranks, rank, rc, i;
  unsigned long *bcast_buff = 0;
  struct timeval t0, t1;

  for (i = 1; i < argc; i++) {
    if (strcmp("--verbose", argv[i]) == 0) {
      verbose = 1;
    } else if (strcmp("--iters", argv[i]) == 0) {
      numiters = atoi(argv[++i]);
    } else if (strcmp("--bcast", argv[i]) == 0) {
      bcast_length = strtoul(argv[++i], 0, 0);
    } else if (strcmp("--reuse", argv[i]) == 0) {
      reuse_buffer = 1;
    } else if (strcmp("--dump-maps", argv[i]) == 0) {
      dump_maps = 1;
    } else {
      printf("xpmem_stress [--bcast <size>] [--iters <N>] [--reuse] [--dump-maps]\n");
      return 0;
    }
  }

  rc = MPI_Init(&argc, &argv);

  if (rc != MPI_SUCCESS) {
    fprintf(stderr, "Error starting MPI program. Terminating.\n");
    goto abort;
  }

  MPI_Comm_size(MPI_COMM_WORLD, &numranks);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  gettimeofday(&t0, NULL);

  for (i = 0; i < numiters; i++) {

    if ((i % 100 == 0) && (rank == 0 || verbose)) {
      printf("[%d] Starting iteration %d / %d\n", rank, i, numiters);
      fflush(stdout);
    }
    if (bcast_length > 0) {

      if (!bcast_buff || !reuse_buffer) {
        bcast_buff = alloc_msg_buffer(bcast_length, rank, rank == 0);
      }

      if (!bcast_buff) {
        rc = -1;
        goto abort;
      }

      rc = MPI_Bcast(bcast_buff, bcast_length / sizeof(unsigned long), MPI_UNSIGNED_LONG, 0, MPI_COMM_WORLD);

      if (rc != MPI_SUCCESS) {
        fprintf(stderr, "(E) [%d] MPI_Bcast failed.\n", rank);
        goto abort;
      }
    }
  }

  gettimeofday(&t1, NULL);

  if (rank == 0 || verbose) {
    printf("[%d] Complete! (time: ~%ld secs)   Finalizing ...\n", rank, t1.tv_sec - t0.tv_sec);
    fflush(stdout);
  }

  if (dump_maps && (rank == 0 || verbose)) {
    dump_self_maps();
  }

  gettimeofday(&t0, NULL);
  MPI_Finalize();
  gettimeofday(&t1, NULL);

  if (rank == 0 || verbose) {
    printf("[%d] Complete! (time: ~%ld secs). Goodbye.\n", rank, t1.tv_sec - t0.tv_sec);
  }

  return 0;

 abort:
  MPI_Abort(MPI_COMM_WORLD, rc);
  return -1;
}


static unsigned long *alloc_msg_buffer(unsigned long size, int rank, int initialize)
{
  unsigned long *buff = (unsigned long *)malloc(size);

  if (!buff) {
    fprintf(stderr, "(E) [%d] Could not allocate message buffer.\n", rank);
    return 0;
  }

  if (initialize) {
    unsigned long i, N;

    for (i = 0, N = size/sizeof(unsigned long); i < N; i++)
      buff[i] = i;
  }

  return buff;
}

static void dump_self_maps(void) {

  char buffer[1024];
  FILE* fp = fopen("/proc/self/maps", "r");

  if (!fp) {
    fprintf(stderr, "(E) could not open /proc/self/maps\n");
    return;
  }

  while (fgets(buffer, 1024, fp))
    printf("%s", buffer);

  fflush(stdout);
}
```

The code is also available at /home/tmusta/public/hangs/xpmem on an Aurora UAN, including a README.   

To see a quick demonstration of the issue, run:
`mpiexec -hostfile -n 12 -ppn 12 --env FI_CXI_DEFAULT_CQ_SIZE=8192 ./xpmem_stress --bcast 32768 --iters 3000`

To simulate when Thunder-SVM is doing:
`mpiexec -hostfile -n 12 -ppn 12 --env FI_CXI_DEFAULT_CQ_SIZE=8192 ./xpmem_stress --bcast 32768 --iters 100000`

To dump the /proc/self/maps file for rank 0, add the "--dump-maps" argument to any.

I am getting a compile fail of 
```
libtool: compile:  icx -DHAVE_CONFIG_H -I. -I./src/include -I./src/include -I./src/util -I./src/mpi/coll/allreduce_group/ -I./src/mpi/coll/transports/gentran -I./src/mpi/coll/transports -I./src/mpi/coll/algorithms/treealgo -I./src/mpi\
/coll/algorithms/recexchalgo -I./src/mpi/coll/algorithms/stubalgo -I./src/mpi/coll/algorithms/common -I./src/mpi/coll/include/ -I./src/mpi/datatype/typerep/src -I./src/mpi/init -I./src/mpi/spawn -I./src/binding/cxx -Isrc/binding/fortr\
an/use_mpi -I./src/mpid/include -I./src/mpid/ch4/include -I./src/mpid/ch4/include -I./src/mpid/ch4/src -I./src/mpid/ch4/netmod/include -I./src/mpid/ch4/netmod/ofi -I./src/mpid/ch4/shm/include -I./src/mpid/ch4/shm/posix -I./src/mpid/ch\
4/shm/src -I./src/mpid/ch4/shm/posix/release_gather -I./src/mpid/ch4/shm/posix/eager/include -I./src/mpid/ch4/shm/posix/eager/include -I./src/mpid/ch4/shm/ipc/src -I./src/mpid/common/sched -I./src/mpid/common/thread -I./src/mpid/commo\
n/timers -I./src/mpid/common/shm -I./src/mpid/common/bc -I./src/mpid/common/genq -I./src/mpid/common/stream_workq -I./src/mpi_t -I/opt/cray/libfabric/1.20.1/include -DNETMOD_INLINE=__netmod_inline_ofi__ -I/lus/flare/projects/Aurora_de\
ployment/bertoni/projects/p01.mpich/7358/src/mpl/include -I/lus/flare/projects/Aurora_deployment/bertoni/projects/p01.mpich/7358/modules/json-c -D_REENTRANT -I/lus/flare/projects/Aurora_deployment/bertoni/projects/p01.mpich/7358/src/mpi/romio/include -I/usr/include -I/lus/flare/projects/Aurora_deployment/bertoni/projects/p01.mpich/7358/src/mpi/datatype/typerep/yaksa/src/frontend/include -fvisibility=hidden -mavx -g -MT src/binding/c/abi/lib_libmpi_la-abi_get_info.\
lo -MD -MP -MF src/binding/c/abi/.deps/lib_libmpi_la-abi_get_info.Tpo -c src/binding/c/abi/abi_get_info.c  -fPIC -DPIC -o src/binding/c/abi/.libs/lib_libmpi_la-abi_get_info.o
icx: remark: Note that use of '-g' without any optimization-level option will turn off most compiler optimizations similar to use of '-O0'; use '-Rno-debug-disables-optimization' to disable this remark [-Rdebug-disables-optimization]
In file included from src/binding/c/abi/abi_get_info.c:8:
In file included from ./src/include/mpiimpl.h:232:
In file included from ./src/mpid/ch4/include/mpidpost.h:10:
In file included from ./src/mpid/ch4/include/mpidch4.h:450:
In file included from ./src/mpid/ch4/shm/src/shm_impl.h:18:
In file included from ./src/mpid/ch4/shm/src/shm_coll.h:10:
In file included from ./src/mpid/ch4/shm/src/../posix/shm_inline.h:16:
In file included from ./src/mpid/ch4/shm/src/../posix/posix_coll.h:12:
In file included from ./src/mpid/ch4/shm/src/../posix/posix_coll_gpu_ipc.h:61:
In file included from ./src/mpid/ch4/shm/src/../posix/../ipc/src/ipc_p2p.h:13:
./src/mpid/ch4/shm/src/../posix/../ipc/src/../xpmem/xpmem_post.h:73:10: error: call to undeclared function 'MPIDI_IPCI_is_repeat_addr'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
   73 |         (MPIDI_IPCI_is_repeat_addr(buf) < 5)) {
      |          ^
1 error generated.
make[2]: *** [Makefile:24523: src/binding/c/abi/lib_libmpi_la-abi_get_info.lo] Error 1
make[2]: Leaving directory '/lus/flare/projects/Aurora_deployment/bertoni/projects/p01.mpich/7358'
make[1]: *** [Makefile:36343: all-recursive] Error 1
make[1]: Leaving directory '/lus/flare/projects/Aurora_deployment/bertoni/projects/p01.mpich/7358'
make: *** [Makefile:6519: all] Error 2
```

The build I used was
```
rm -rf 7358
git clone git@github.com:hzhou/mpich.git 7358
rm -rf install_7358
mkdir install_7358
cd 7358
git checkout 2503_repeat_addr
git submodule update --init
./autogen.sh
./configure --prefix=$PWD/../install_7358 --disable-maintainer-mode --disable-silent-rules --enable-shared --with-pm=no --enable-romio --without-ibverbs --enable-wrapper-rpath=yes   --with-slurm=no --with-pmi=pmix --with-pmix=/usr   -\
-without-cuda --without-hip --with-device=ch4:ofi --with-libfabric=/opt/cray/libfabric/1.20.1 --enable-libxml2 --with-datatype-engine=yaksa --with-xpmem=/usr --with-file-system=daos+lustre+nfs+ufs --with-daos=/usr --enable-timer-type=linux86_cycle --enable-fast=avx --enable-g=none --disable-debuginfo  --without-valgrind --enable-ch4-mt=runtime --with-ze --disable-opencl --enable-g=all
```

Any ideas? Or did I miss something?

--- a/src/mpid/ch4/shm/ipc/gpu/gpu_post.c
+++ b/src/mpid/ch4/shm/ipc/gpu/gpu_post.c
@@ -374,8 +374,9 @@ int MPIDI_GPU_get_ipc_attr(const void *buf, MPI_Aint count, MPI_Datatype datatyp
     MPIR_ERR_CHKANDJUMP(mpl_err != MPL_SUCCESS, mpi_errno, MPI_ERR_OTHER, "**gpu_get_buffer_info");
 
     /* Don't create or search for an ipc handle if the buffer doesn't meet the threshold
-     * requirement */
-    if (bounds_len < MPIR_CVAR_CH4_IPC_GPU_P2P_THRESHOLD && !MPIDI_IPCI_is_repeat_addr(bounds_base)) {
+     * requirement or it isn't a repeat address */
+    if (bounds_len < MPIR_CVAR_CH4_IPC_GPU_P2P_THRESHOLD &&
+        (MPIDI_IPCI_is_repeat_addr(bounds_base) < 1)) {
         goto fn_exit;
     }
 

--- a/src/mpid/ch4/shm/ipc/src/globals.c
+++ b/src/mpid/ch4/shm/ipc/src/globals.c
@@ -6,8 +6,44 @@
 #include "mpidimpl.h"
 #include "ipc_types.h"
 
+/*
+=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
+cvars:
+    - name        : MPIR_CVAR_CH4_IPC_MAP_REPEAT_ADDR
+      category    : CH4
+      type        : boolean
+      default     : true
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        If an address is used more than once in the last ten send operations,
+        map it for IPC use even if it is below the IPC threshold.
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
 MPIDI_IPCI_global_t MPIDI_IPCI_global;
 
 #ifdef MPL_USE_DBG_LOGGING
 MPL_dbg_class MPIDI_IPCI_DBG_GENERAL;
 #endif
+
+bool MPIDI_IPCI_is_repeat_addr(void *addr)
+{
+    if (!MPIR_CVAR_CH4_IPC_MAP_REPEAT_ADDR) {
+        return false;
+    }
+
+    static void *repeat_addr[10] = { 0 };
+    static int addr_idx = 0;
+
+    for (int i = 0; i < 10; i++) {
+        if (addr == repeat_addr[i]) {
+            return true;
+        }
+    }
+
+    repeat_addr[addr_idx] = addr;
+    addr_idx = (addr_idx + 1) % 10;
+    return false;
+}

--- a/src/mpid/ch4/shm/ipc/src/globals.c
+++ b/src/mpid/ch4/shm/ipc/src/globals.c
@@ -17,8 +17,15 @@ cvars:
       verbosity   : MPI_T_VERBOSITY_USER_BASIC
       scope       : MPI_T_SCOPE_ALL_EQ
       description : >-
-        If an address is used more than once in the last ten send operations,
-        map it for IPC use even if it is below the IPC threshold.
+        Enable to track how often a buffer is being sent repeatedly. This will
+        be used in determine whether to use IPC algorithm to deliver the message.
+        The choice will depend on the IPC driver. In the case of high-latency
+        buffers such as GPU device buffer, we will enable IPC if EITHER the
+        message size is above a threshold or the message buffer is being repeated.
+        On the other hand, if the address mapping overhead is relatively high,
+        such as the case for XPMEM, we will enable IPC when BOTH conditions --
+        message size and repeat count -- are met.
+
 === END_MPI_T_CVAR_INFO_BLOCK ===
 */
 
@@ -28,22 +35,25 @@ MPIDI_IPCI_global_t MPIDI_IPCI_global;
 MPL_dbg_class MPIDI_IPCI_DBG_GENERAL;
 #endif
 
-bool MPIDI_IPCI_is_repeat_addr(void *addr)
+int MPIDI_IPCI_is_repeat_addr(const void *addr)
 {
     if (!MPIR_CVAR_CH4_IPC_MAP_REPEAT_ADDR) {
-        return false;
+        return 0;
     }
-
-    static void *repeat_addr[10] = { 0 };
+#define MAX_REPEAT_ADDRS 16     /* a bit arbitrary - TODO: make it tunable */
+    static const void *repeat_addr[MAX_REPEAT_ADDRS] = { 0 };
+    static int repeat_addr_count[MAX_REPEAT_ADDRS];
     static int addr_idx = 0;
 
-    for (int i = 0; i < 10; i++) {
+    for (int i = 0; i < MAX_REPEAT_ADDRS; i++) {
         if (addr == repeat_addr[i]) {
-            return true;
+            repeat_addr_count[i]++;
+            return repeat_addr_count[i];
         }
     }
 
     repeat_addr[addr_idx] = addr;
-    addr_idx = (addr_idx + 1) % 10;
-    return false;
+    repeat_addr_count[addr_idx] = 0;
+    addr_idx = (addr_idx + 1) % MAX_REPEAT_ADDRS;
+    return 0;
 }

--- a/src/mpid/ch4/shm/ipc/src/ipc_send.h
+++ b/src/mpid/ch4/shm/ipc/src/ipc_send.h
@@ -13,42 +13,6 @@
 #include "../gpu/gpu_post.h"
 #include "ipc_p2p.h"
 
-/*
-=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
-cvars:
-    - name        : MPIR_CVAR_CH4_IPC_MAP_REPEAT_ADDR
-      category    : CH4
-      type        : boolean
-      default     : true
-      class       : none
-      verbosity   : MPI_T_VERBOSITY_USER_BASIC
-      scope       : MPI_T_SCOPE_ALL_EQ
-      description : >-
-        If an address is used more than once in the last ten send operations,
-        map it for IPC use even if it is below the IPC threshold.
-=== END_MPI_T_CVAR_INFO_BLOCK ===
-*/
-
-MPL_STATIC_INLINE_PREFIX bool MPIDI_IPCI_is_repeat_addr(void *addr)
-{
-    if (!MPIR_CVAR_CH4_IPC_MAP_REPEAT_ADDR) {
-        return false;
-    }
-
-    static void *repeat_addr[10] = { 0 };
-    static int addr_idx = 0;
-
-    for (int i = 0; i < 10; i++) {
-        if (addr == repeat_addr[i]) {
-            return true;
-        }
-    }
-
-    repeat_addr[addr_idx] = addr;
-    addr_idx = (addr_idx + 1) % 10;
-    return false;
-}
-
 MPL_STATIC_INLINE_PREFIX int MPIDI_IPCI_try_lmt_isend(const void *buf, MPI_Aint count,
                                                       MPI_Datatype datatype, int rank, int tag,
                                                       MPIR_Comm * comm, int attr,

--- a/src/mpid/ch4/shm/ipc/src/ipc_types.h
+++ b/src/mpid/ch4/shm/ipc/src/ipc_types.h
@@ -61,6 +61,6 @@ extern MPL_dbg_class MPIDI_IPCI_DBG_GENERAL;
 
 extern MPIDI_IPCI_global_t MPIDI_IPCI_global;
 
-bool MPIDI_IPCI_is_repeat_addr(void *addr);
+int MPIDI_IPCI_is_repeat_addr(const void *addr);
 
 #endif /* IPC_TYPES_H_INCLUDED */

--- a/src/mpid/ch4/shm/ipc/src/ipc_types.h
+++ b/src/mpid/ch4/shm/ipc/src/ipc_types.h
@@ -61,4 +61,6 @@ extern MPL_dbg_class MPIDI_IPCI_DBG_GENERAL;
 
 extern MPIDI_IPCI_global_t MPIDI_IPCI_global;
 
+bool MPIDI_IPCI_is_repeat_addr(void *addr);
+
 #endif /* IPC_TYPES_H_INCLUDED */

--- a/src/mpid/ch4/shm/ipc/xpmem/xpmem_post.h
+++ b/src/mpid/ch4/shm/ipc/xpmem/xpmem_post.h
@@ -69,7 +69,8 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_get_ipc_attr(const void *buf, MPI_Aint 
     if (!MPIR_CVAR_CH4_XPMEM_ENABLE || buf == MPI_BOTTOM ||
         data_sz < MPIR_CVAR_CH4_IPC_XPMEM_P2P_THRESHOLD ||
         (MPIR_CVAR_CH4_IPC_XPMEM_P2P_UPPER_THRESHOLD > 0 &&
-         data_sz > MPIR_CVAR_CH4_IPC_XPMEM_P2P_UPPER_THRESHOLD)) {
+         data_sz > MPIR_CVAR_CH4_IPC_XPMEM_P2P_UPPER_THRESHOLD) ||
+        (MPIDI_IPCI_is_repeat_addr(buf) < 5)) {
         goto fn_exit;
     } else {
         ipc_attr->ipc_type = MPIDI_IPCI_TYPE__XPMEM;

